x-pack/plugins/security_solution/public/detections/components/alerts_table/timeline_actions/alert_context_menu.tsx

```diff
@@ -0,0 +1,457 @@
+import { useDispatch } from 'react-redux';
+import {
+  EuiText,
+  EuiButtonIcon,
+  EuiContextMenuPanel,
+  EuiPopover,
+  EuiContextMenuItem,
+} from '@elastic/eui';
+import styled from 'styled-components';
+
+import { DEFAULT_INDEX_PATTERN } from '../../../../../common/constants';
+import { Status } from '../../../../../common/detection_engine/schemas/common/schemas';
+import { isThresholdRule } from '../../../../../common/detection_engine/utils';
+import { RuleType } from '../../../../../common/detection_engine/types';
+import { isMlRule } from '../../../../../common/machine_learning/helpers';
+import { timelineActions } from '../../../../timelines/store/timeline';
+import { EventsTd, EventsTdContent } from '../../../../timelines/components/timeline/styles';
+import { DEFAULT_ICON_BUTTON_WIDTH } from '../../../../timelines/components/timeline/helpers';
+import { FILTER_OPEN, FILTER_CLOSED, FILTER_IN_PROGRESS } from '../alerts_filter_group';
+import { updateAlertStatusAction } from '../actions';
+import { SetEventsDeletedProps, SetEventsLoadingProps } from '../types';
+import { Ecs, TimelineNonEcsData } from '../../../../graphql/types';
+import {
+  AddExceptionModal as AddExceptionModalComponent,
+  AddExceptionModalBaseProps,
+} from '../../../../common/components/exceptions/add_exception_modal';
+import { getMappedNonEcsValue } from '../../../../common/components/exceptions/helpers';
+import * as i18n from '../translations';
+import {
+  useStateToaster,
+  displaySuccessToast,
+  displayErrorToast,
+} from '../../../../common/components/toasters';
+import { useUserData } from '../../user_info';
+
+interface AlertContextMenuProps {
+  ecsRowData: Ecs;
+  nonEcsRowData: TimelineNonEcsData[];
+  timelineId: string;
+}
+
+const addExceptionModalInitialState: AddExceptionModalBaseProps = {
+  ruleName: '',
+  ruleId: '',
+  ruleIndices: [],
+  exceptionListType: 'detection',
+  alertData: undefined,
+};
+
+export const AlertContextMenu = React.memo<AlertContextMenuProps>(
+  ({ ecsRowData, nonEcsRowData, timelineId }) => {
+    const dispatch = useDispatch();
+    const [, dispatchToaster] = useStateToaster();
+    const [isPopoverOpen, setPopover] = useState(false);
+    const alertStatus =
+      (ecsRowData.signal?.status && (ecsRowData.signal.status[0] as Status)) ?? undefined;
+    const eventId = ecsRowData._id;
+
+    const onButtonClick = useCallback(() => {
+      setPopover(!isPopoverOpen);
+    }, [isPopoverOpen]);
+
+    const closePopover = useCallback(() => {
+      setPopover(false);
+    }, []);
+    const [shouldShowAddExceptionModal, setShouldShowAddExceptionModal] = useState(false);
+    const [addExceptionModalState, setAddExceptionModalState] = useState<
+      AddExceptionModalBaseProps
+    >(addExceptionModalInitialState);
+    const [{ canUserCRUD, hasIndexWrite }] = useUserData();
+
+    const isEndpointAlert = useMemo(() => {
+      if (!nonEcsRowData) return false;
```

Comment:
nit: although it's not called out in the [STYLEGUIDE](https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md), there doesn't appear to be many (any?) instances of early-returning via a one liner. If you agree, consider adopting the prevailing style.